libcodechecker/report.py

```diff
@@ -0,0 +1,444 @@
+        self.col = col
+        self.file_path = filepath
+
+    def __str__(self):
+        msg = "@"+str(self.line)+":"+str(self.col)+"::"+str(self.file_path)
+        return msg
+
+
+class Range(object):
+    """
+    Two positions build a range.
+    """
+    def __init__(self, begin, end=None):
+        self.__begin = begin
+        if end:
+            self.__end = end
+        else:
+            self.__end = begin
+
+    def __str__(self):
+        msg = "["+str(self.__begin) + "][" + str(self.__end)+"]"
+        return msg
+
+    @property
+    def begin(self):
+        return self.__begin
+
+    @property
+    def end(self):
+        return self.__end
+
+
+class BugPath(object):
+    """
+    Represents a bug path which contains multiple DiagSections.
+    The order of the diagnostic sections in the bug path matters, and
+    needed for the storage in the database and for the UI to
+    render arrows properly.
+    """
+    def __init__(self):
+        self.__diag_sections = []
+        self.__path_length = 0
+
+    def add_diag(self, diag_section):
+        """
+        The order of the diagnostic sections in the bug path matters,
+        needed for the UI to render arrows properly.
+        """
+        diag_section.path_position = self.__path_length
+        self.__diag_sections.append(diag_section)
+        self.__path_length += 1
+
+    def get_diag_sections(self):
+        return self.__diag_sections
+
+    def len(self):
+        """
+        Contains all the control and event diag sections!
+        """
+        return self.__path_length
+
+    @property
+    def path_hash(self):
+        """
+        Generate a unique id for a bug path.
+        """
+        # TODO: Review hash generation method this is just an initial version.
+        # Hash is generated now based on columm numbers from the diag sections
+        # in a bug path.
+        # At least one event should be in the bug path.
+        # NOTE: Events might only have only a location and
+        # no range information.
+        path = ""
+        for ds in self.__diag_sections:
+            path += str(ds.start_range.begin.col) + \
+                str(ds.end_range.begin.col) + \
+                ds.msg + ds.start_range.begin.file_path
+
+        return hashlib.sha1(path).hexdigest()
+
+
+class DiagSection(GenericEquality):
+    """
+    Diagnostic section can have a start and an end position.
+    If position_in_path is set this section belongs to a BugPath.
+    It is possible that a DiagSection is a standalone "message" or
+    "warning" by the analyzer and does not belong to BugPath.
+
+    See storage thrift api desctiption for kinds.
+    """
+
+    def __init__(self, kind='', msg=''):
+
+        self.__start_range = None
+        self.__end_range = None
+        self.__msg = msg
+        self.__extended_msg = ''
+        self.__kind = kind
+
+        # One Position.
+        self.__location = None  # Optional in some cases only range is
```

Comment:
Maybe it would be worth to have a catalog of these diagnostic pieces, like a list of example plist fragments about the cases that we would like to cover.